### PR TITLE
Tenant cordoning

### LINKED ISSUE #243: Cordon a tenant for maintenance
# Describe the feature
Inspired by  the [kube-anchor](https://github.com/JESWINKNINAN/kube-anchor) project, it would be nice to leave the cluster admin to "cordon" a tenant for maintenance during a Production Freeze Window.

### Some possible use cases:

- Avoid accidental resource modification(s) including deletion during a Production Freeze Window
- During Kubernetes upgrade, to prevent any workload updates
- During incidents or outages
- During planned maintenance of a dedicated nodes pool in a BYOD scenario 

# What would the new user story look like?
1. The cluster admin has to schedule a Production Freeze Window for maintenance
2. He informs the tenant owners that their tenants will be frozen for a while. During this window, no CREATE/UPDATE/DELETE operations are permitted in the impacted tenants.
3. The cluster admin places a label to the tenant, eg.
               
```
$ kubectl label tenant oil capsule.clastix.io/cordon=enabled
tenant oil labeled
```

# Expected behavior
When a tenant is cordoned by cluster admin, a Capsule Admission WebHook denies all CREATE/UPDATE/DELETE user's requests in the cordoned tenant's namespaces. A meaningful message should be displayed to the tenant owner informing that during the maintenance window, no CREATE/UPDATE/DELETE operations are permitted.

### PR DESCRIPTION
**Pull request by prometherion**
Closes #243.

### CODE REVIEW

**Review by bsctl — COMMENTED** [review]
@prometherion
A tenant owner can still create namespaces in a cordoned tenant.

**Comment by bsctl** [comment]
@prometherion 
As further improvement, we should keep track of the cordoning/uncordoning events, es.

```
$ kubectl describe tnt oil
...

Events:
  Type    Reason  Age  From  Message
  Warning  TenantCordoned  21m  tenant-webhook  tenant oil is cordoned by cluster admin
  Warning  TenantUnCordoned  21m  tenant-webhook  tenant oil is uncordoned by cluster admin
```

**Comment by prometherion** [comment]
> As further improvement, we should keep track of the cordoning/uncordoning events, es.

Unfortunately, we cannot perform these events.

But we can track violations (sic) of users trying to create, update, or delete resources in the Namespaces.

**Comment by prometherion** [comment]
Along with this, we're tracking also events

```
  Type     Reason                    Age                      From               Message
  ----     ------                    ----                     ----               -------

  Warning  TenantFreezed             7s                       tenant-webhook     apps/v1, Kind=Deployment oil-development/nginx cannot be createed, current Tenant is freezer
````

**Review by bsctl — COMMENTED** [review]
@prometherion 
With latest commit, after cordoning a tenant, the namespace creation is blocked but the creation of resources, es. pods, in the other namespaces is not blocked by the webhook. Pls, check it.

**Comment by prometherion** [comment]
Everything looks good to me, but I guess you missed the part where a cluster administrator request is not going to be blocked by the validation webhook.

When a Tenant is cordoned, every request made by a Tenant Owner or SA in the Namespaces managed by the Tenant is going to be rejected.

Said so, if a cluster administrator starts a Pod in a cordoned Namespace, this will not be rejected: this is mandatory since we have some components that must be able to perform actions also on freezed Namespaces, as the Kubernetes Controller Managers or the cluster administrator itself.

**Comment by bsctl** [comment]
> Everything looks good to me

@prometherion The issue was related to a missing `validatingwebhookconfiguration`. Actually my bad.

**Comment by bsctl** [comment]
@prometherion

It seems there's another issue.

* This time, it started from scratch

```
~/capsule$ make remove
```

```
~/capsule$ make docker-build
```

```
~/capsule$ make install
go: creating new go.mod: module tmp
go get: added sigs.k8s.io/controller-tools v0.5.0
/home/adriano/go/bin/controller-gen "crd:trivialVersions=true,preserveUnknownFields=false" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
go: creating new go.mod: module tmp
go get: added sigs.k8s.io/kustomize/kustomize/v3 v3.5.4
/home/adriano/go/bin/kustomize build config/crd | kubectl apply -f -
customresourcedefinition.apiextensions.k8s.io/capsuleconfigurations.capsule.clastix.io created
customresourcedefinition.apiextensions.k8s.io/tenants.capsule.clastix.io created
```

```
~/capsule$ make deploy
go: creating new go.mod: module tmp
go get: added sigs.k8s.io/controller-tools v0.5.0
/home/adriano/go/bin/controller-gen "crd:trivialVersions=true,preserveUnknownFields=false" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
go: creating new go.mod: module tmp
go get: added sigs.k8s.io/kustomize/kustomize/v3 v3.5.4
cd config/manager && /home/adriano/go/bin/kustomize edit set image controller=quay.io/clastix/capsule:$(git describe --abbrev=0 --tags)
/home/adriano/go/bin/kustomize build config/default | kubectl apply -f -
namespace/capsule-system unchanged
customresourcedefinition.apiextensions.k8s.io/capsuleconfigurations.capsule.clastix.io configured
customresourcedefinition.apiextensions.k8s.io/tenants.capsule.clastix.io configured
mutatingwebhookconfiguration.admissionregistration.k8s.io/capsule-mutating-webhook-configuration created
clusterrolebinding.rbac.authorization.k8s.io/capsule-manager-rolebinding created
secret/capsule-ca created
secret/capsule-tls created
service/capsule-controller-manager-metrics-service created
service/capsule-webhook-service created
deployment.apps/capsule-controller-manager created
capsuleconfiguration.capsule.clastix.io/capsule-default created
validatingwebhookconfiguration.admissionregistration.k8s.io/capsule-validating-webhook-configuration created
```

* capsule pod is not starting

```
$ kubectl get pods
NAME                                          READY   STATUS              RESTARTS   AGE
capsule-controller-manager-7c674c46dd-gj4hb   0/1     ContainerCreating   0          29m
```

* check kube-apiserver logs

```
$ sudo crictl ps | grep apiserver
acfb7041873c2       d7e24aeb3b102       2 hours ago         Running             kube-apiserver            12                  662d707a27ce8
```

```
~$ sudo crictl logs acfb7041873c2
...
W0622 17:56:35.186575       1 dispatcher.go:134] Failed calling webhook, failing closed cordoning.tenant.capsule.clastix.io: failed calling webhook "cordoning.tenant.capsule.clastix.io": Post "https://capsule-webhook-service.capsule-system.svc:443/tenant-cordoning?timeout=10s": dial tcp 10.32.121.87:443: connect: connection refused
W0622 17:56:35.189054       1 dispatcher.go:134] Failed calling webhook, failing closed cordoning.tenant.capsule.clastix.io: failed calling webhook "cordoning.tenant.capsule.clastix.io": Post "https://capsule-webhook-service.capsule-system.svc:443/tenant-cordoning?timeout=10s": dial tcp 10.32.121.87:443: connect: connection refused

```

* it looks the `validatingwebhookconfiguration` blocks the pods creation

```yaml
- admissionReviewVersions:
  - v1
  clientConfig:
    service:
      name: capsule-webhook-service
      namespace: capsule-system
      path: /tenant-cordoning
      port: 443
  failurePolicy: Fail                                      # <<< here
  matchPolicy: Equivalent
  name: cordoning.tenant.capsule.clastix.io
  namespaceSelector: {}
  objectSelector: {}
  rules:
  - apiGroups:
    - '*'
    apiVersions:
    - '*'
    operations:
    - CREATE
    - UPDATE
    - DELETE
    resources:
    - '*'
    scope: '*'
  sideEffects: None
  timeoutSeconds: 10
```

* by changing `failurePolicy` from `Fail` to `Ignore` , the capsule pods gets created

```
kubectl get pods
NAME                                          READY   STATUS    RESTARTS   AGE
capsule-controller-manager-7c674c46dd-gj4hb   1/1     Running   0          48m
```

**Comment by prometherion** [comment]
I got it, we're missing the `namespaceSelector` on that webhook.

File: pkg/webhook/namespace/freezed.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+		for _, objectRef := range ns.ObjectMeta.OwnerReferences {
+			// retrieving the selected Tenant
+			tnt := &capsulev1alpha1.Tenant{}
+			if err := client.Get(ctx, types.NamespacedName{Name: objectRef.Name}, tnt); err != nil {
+				return admission.Errored(http.StatusBadRequest, err)
+			}
+
+			if tnt.IsCordoned() {
+				recorder.Eventf(tnt, corev1.EventTypeWarning, "TenantFreezed", "Namespaced %s cannot be attached, the current Tenant is freezed", ns.GetName())
```
Comment (bsctl) [inline pkg/webhook/namespace/freezed.go:70]: use _Namespace_ instead of _Namespaced_

File: pkg/webhook/tenant/cordoning.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+	// resource is not inside a Tenant namespace
+	if len(tntList.Items) == 0 {
+		return admission.Allowed("")
+	}
+
+	tnt := tntList.Items[0]
+
+	if tnt.IsCordoned() {
+		if utils.RequestFromOwnerOrSA(tnt, req, h.configuration.UserGroups()) {
+			recorder.Eventf(&tnt, corev1.EventTypeWarning, "TenantFreezed", "%s %s/%s cannot be %sed, current Tenant is freezed", req.Kind.String(), req.Namespace, req.Name, strings.ToLower(string(req.Operation)))
```
Comment (bsctl) [inline pkg/webhook/tenant/cordoning.go:73]: small typo:

`apps/v1, Kind=Deployment oil-production/example cannot be **updateed**, current Tenant is freezed`
`/v1, Kind=Pod oil-development/example cannot be **createed**, current Tenant is freezed`

**Review by bsctl — CHANGES_REQUESTED** [review]
@prometherion just few typo in messages. A part of that, LGTM.

**Comment by prometherion** [comment]
Thanks @bsctl, just fixed the typos.

**Review by bsctl — APPROVED** [review]

**Pull request merged**